crates/shared/src/price_estimation/buffered.rs

```diff
@@ -0,0 +1,459 @@
+{
+    /// Maximum capacity of the broadcast channel, the messages are discarded as
+    /// soon as they are sent, so this limit should be enough to hold the
+    /// flow
+    const BROADCAST_CHANNEL_CAPACITY: usize = 50;
+
+    /// Creates a new buffered transport with the specified configuration.
+    pub fn with_config(inner: Inner, config: Configuration) -> Self {
+        let inner = Arc::new(inner);
+        let (calls, receiver) = mpsc::unbounded();
+
+        let (broadcast_sender, _) = broadcast::channel(Self::BROADCAST_CHANNEL_CAPACITY);
+
+        Self::background_worker(
+            inner.clone(),
+            config.clone(),
+            receiver,
+            broadcast_sender.clone(),
+        );
+
+        Self {
+            inner,
+            calls,
+            broadcast_sender,
+            config,
+        }
+    }
+
+    /// Start a background worker for handling batched requests.
+    fn background_worker(
+        inner: Arc<Inner>,
+        config: Configuration,
+        calls: mpsc::UnboundedReceiver<H160>,
+        broadcast_sender: broadcast::Sender<NativePriceResult>,
+    ) -> JoinHandle<()> {
+        tokio::task::spawn(batched_for_each(config, calls, move |batch| {
+            let inner = inner.clone();
+            let broadcast_sender = broadcast_sender.clone();
+            async move {
+                let batch = batch.into_iter().collect::<HashSet<_>>();
+                if batch.len() != 0 {
+                    let results = match inner.fetch_native_prices(&batch).await {
+                        Ok(results) => results
+                            .into_iter()
+                            .map(|(token, price)| (token, Ok::<_, PriceEstimationError>(price)))
+                            .collect::<HashMap<_, _>>(),
+                        Err(err) => {
+                            tracing::error!(?err, "failed to send native price batch request");
+                            batch
+                                .into_iter()
+                                .map(|token| (token, Err(err.clone())))
+                                .collect::<HashMap<_, _>>()
+                        }
+                    };
+                    for result in results {
+                        let _ = broadcast_sender.send(result);
+                    }
+                }
+            }
+        }))
+    }
+
+    /// Blocking operation to get estimate prices in a batch
+    pub async fn blocking_buffered_estimate_prices(
+        &self,
+        token: &H160,
+    ) -> NativePriceEstimateResult {
+        // Sends the token for requesting price
+        self.calls.unbounded_send(*token).map_err(|e| {
+            PriceEstimationError::ProtocolInternal(anyhow!(
+                "failed to append a new token to the queue: {e:?}"
+            ))
+        })?;
+
+        let mut rx = self.broadcast_sender.subscribe();
+
+        tokio::time::timeout(self.config.batch_delay.saturating_mul(3), async {
```

Comment:
We subscribe to the broadcast and wait for our result 3 times the `batch_delay`. I thought 3 is a compromise. I don't want to set too short time, otherwise we may lose valid results, but I don't want to put too high either, because in case of no valid result, not to be blocked for so much time.

Reply:
Does it make sense to extract this timeout param from the config?

Reply:
This magic number is weird. I think this method should just take as long as it does and if someone wants to timeout earlier they should combine it with a timeout future.